src/common/eeprom.c

```diff
@@ -2,230 +2,290 @@
 
 #include "eeprom.h"
 #include <stdio.h>
+#include <string.h>
+#include "config.h"
 #include "st25dv64k.h"
 #include "dbg.h"
-#include <string.h>
-#include <stdbool.h>
-
-#define EE_VERSION 0x0003
-#define EE_VAR_CNT sizeof(eeprom_map_v1)
-#define EE_ADDRESS 0x0500
-
-const uint8_t eeprom_map_v1[] = {
-    VARIANT8_UI16, // EEVAR_VERSION
-    VARIANT8_UI8,  // EEVAR_FILAMENT_TYPE
-    VARIANT8_UI32, // EEVAR_FILAMENT_COLOR
-    VARIANT8_FLT,  //
-    VARIANT8_FLT,  //
-    VARIANT8_FLT,  //
-    VARIANT8_UI8,  // EEVAR_RUN_SELFTEST
-    VARIANT8_UI8,  // EEVAR_RUN_XYZCALIB
-    VARIANT8_UI8,  // EEVAR_RUN_FIRSTLAY
-    VARIANT8_UI8,  // EEVAR_FSENSOR_ENABLED
-    VARIANT8_UI8,  // EEVAR_LAN_FLAG
-    VARIANT8_UI32, // EEVAR_LAN_IP4_ADDR    X.X.X.X address encoded in uint32
-    VARIANT8_UI32, // EEVAR_LAN_IP4_MSK     X.X.X.X address encoded in uint32
-    VARIANT8_UI32, // EEVAR_LAN_IP4_GW      X.X.X.X address encoded in uint32
-    VARIANT8_UI32, // EEVAR_LAN_IP4_DNS1    X.X.X.X address encoded in uint32
-    VARIANT8_UI32, // EEVAR_LAN_IP4_DNS2    X.X.X.X address encoded in uint32
-    VARIANT8_UI8,  // EEVAR_LAN_HOSTNAME_START   Start of 20char string
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
+
+
+#define EEPROM_VARCOUNT  (sizeof(eeprom_map) / sizeof(eeprom_entry_t))
+#define EEPROM_DATASIZE sizeof(eeprom_vars_t)
+
+
+#pragma pack(push)
+#pragma pack(1)
+
+// eeprom map entry structure
+typedef struct _eeprom_entry_t {
+	uint8_t type;   // variant8 data type
+	uint8_t count;  // number of elements
+} eeprom_entry_t;
+
+// eeprom vars structure (used for defaults)
+typedef struct _eeprom_vars_t {
+    uint16_t VERSION;
+    uint16_t FEATURES;
+    uint16_t DATASIZE;
+    uint16_t FWVERSION;
+    uint16_t FWBUILD;
+    uint8_t FILAMENT_TYPE;
+    uint32_t FILAMENT_COLOR;
+    uint8_t RUN_SELFTEST;
+    uint8_t RUN_XYZCALIB;
+    uint8_t RUN_FIRSTLAY;
+    uint8_t FSENSOR_ENABLED;
+    float ZOFFSET;
+    float PID_NOZ_P;
+    float PID_NOZ_I;
+    float PID_NOZ_D;
+    float PID_BED_P;
+    float PID_BED_I;
+    float PID_BED_D;
+    uint8_t LAN_FLAG;
+    uint32_t LAN_IP4_ADDR;
+	uint32_t LAN_IP4_MSK;
+	uint32_t LAN_IP4_GW;
+	uint32_t LAN_IP4_DNS1;
+	uint32_t LAN_IP4_DNS2;
+	char LAN_HOSTNAME[LAN_HOSTNAME_MAX_LEN + 1];
+    char TEST[10];
+    uint32_t CRC32;
```

Comment:
the CRC is the same case as a version number - it should probably be among the first structure items to ensure fixed location for future expansion of the EEPROM struct

Reply:
CRC is not same case. There is no reason to place CRC into "header" at start of the structure other than estetic. We can do version independent check with crc at end also with more simple code because we do not need skip it and calculation is one single call.